### PR TITLE
gh-96821: Fix undefined behaviour in `_testcapimodule.c`

### PR DESCRIPTION
**Pull request by matthiasgoergens**
The first commit demonstrates the undefined behaviour and is meant to fail tests.

The other commits fix it.

* Issue: gh-96821

### CODE REVIEW

**Comment by matthiasgoergens** [comment]
@kumaraditya303 @mdickinson Would you please have a look?

File: Misc/NEWS.d/next/Core and Builtins/2022-09-18-08-47-40.gh-issue-96821.Co2iOq.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Fix undefined behaviour in `_testcapimodule.c`.
```
Comment (CAM-Gerlach) [inline Misc/NEWS.d/next/Core and Builtins/2022-09-18-08-47-40.gh-issue-96821.Co2iOq.rst]:
```suggestion
Fix undefined behaviour in ``_testcapimodule.c``.
```

I assume you mean to use a literal here, not the default role (which renders as italic prose).

Comment (matthiasgoergens) [reply Misc/NEWS.d/next/Core and Builtins/2022-09-18-08-47-40.gh-issue-96821.Co2iOq.rst]: Thanks!

**Review by Fidget-Spinner — APPROVED** [review]
I think I've wrote similar code before elsewhere (just doing raw pointer arithmetic on an unverified args pointer). However, I don't recall where exactly now :(.

Anyways, thanks this LGTM.

**Comment by matthiasgoergens** [comment]
Thanks!

**Comment by matthiasgoergens** [comment]
> I think I've wrote similar code before elsewhere (just doing raw pointer arithmetic on an unverified args pointer). However, I don't recall where exactly now :(.

There was an example of this in eg `ceval.c`. Not sure if that one was yours?

Btw, this kind of null pointer arithmetic is perfectly defined, as long as we compile with `-fwrapv`. I'm just working on cleaning things up so we can get rid of that flag and enjoy a few more C compiler optimisations.

**Pull request merged**